### PR TITLE
[Android/iOS Visual] added disabled text button color effect

### LINKED ISSUE #5399: Disabled Button Appearance
### Description

**App.xaml**
```
<Style x:Key="DefaultButtonStyle" TargetType="Button">
    <Setter Property="VerticalOptions" Value="Center" />
    <Setter Property="BackgroundColor" Value="Green" />
    <Setter Property="CornerRadius" Value="10" />
    <Setter Property="TextColor" Value="White" />
</Style>
```

**View.xaml**
```
<Button
    Style="{x:Static res:Styles.DefaultButtonStyle}"
    IsEnabled="false"
    Text="Sign" />
```

From the code above, it produces an undesirable look of a `Button`. The result would either be: text is not visible or the default `IsEnabled` appearance doesn't get applied. It still retain the other properties such as button click event is disabled.

This bug only occurs in Xamarin.iOS

### Steps to Reproduce

1. Copy-paste the code provided above into a working project.
2. Debug it into an iPhone Simulator or an actual iPhone device.
3. Navigate to the page where the button is located.
4. Notice the appearance of the button.

### Expected Behavior
The button should appear disabled while having a corner radius, green background and white text color.

### Actual Behavior
Text is not visible or the default IsEnabled appearance doesn't get applied

### Basic Information

- Version with issue: 3.5.0.129452
- Last known good version: NA
- IDE: Visual Studio for Mac
- Platform Target Frameworks:
  - iOS:  8.0
  - Android: NA
  - UWP:  NA
- Android Support Library Version: NA
- Nuget Packages: Prism.DryIoc.Forms, Toast.iOS, Xamarin.FFImageLoading, Xamarin.Forms
- Affected Devices: iPhone 7 Plus - 2, iPhone XS Max 12.1 Simulator

### Screenshots

Will get back to this

### Reproduction Link

Will get back to this

### PR DESCRIPTION
**Pull request by emmanuelvanmeerbeeck3factr**
### Description of Change ###

Added a `DisabledTextColor` property to the `Button` - element. This property can be used to change the text color of a `Button` in its disabled state.

### Issues Resolved ### 
- fixes #5399

### API Changes ###

Added:
 - `Xamarin.Forms.Color` Button.DisabledTextColor { get; set; } //Bindable Property

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android

### Behavioral/Visual Changes ###
A user can use this bindable property to change the text color of the button in its disabled state.

### Before/After Screenshots ### 
BEFORE
simulator screen shot - iphone 6 - 2019-02-18 at 13 16 22
AFTER (With `DisabledTextColor` set on `Green`)
simulator screen shot - iphone 6 - 2019-02-18 at 13 18 28

### Testing Procedure ###
Add the `DisabledTextColor` to a `Button` and assign it a sample `Xamarin.Forms.Color` - value.

### PR Checklist ###

- [ ] Has automated tests
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

###Notes###
Contributions made by the entire **3factr** - team.

### CODE REVIEW

File: Xamarin.Forms.Core/Button.cs [hunk 1]
```diff
@@ -52,6 +52,8 @@ public class Button : View, IFontElement, ITextElement, IBorderElement, IButtonC
 
 		public static readonly BindableProperty PaddingProperty = PaddingElement.PaddingProperty;
 
+		public static readonly BindableProperty DisabledTextColorProperty = BindableProperty.Create("DisabledTextColor", typeof(Color), typeof(Button), Color.Default);
```
Comment (paymicro) [inline Xamarin.Forms.Core/Button.cs]:
```suggestion
		public static readonly BindableProperty DisabledTextColorProperty = BindableProperty.Create(nameof(DisabledTextColor), typeof(Color), typeof(Button), Color.Default);
```

File: Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs [hunk 2]
```diff
@@ -182,6 +187,16 @@ void UpdateTextColor()
...
 		}
 
+		void UpdateDisabledTextColor()
+		{
+			int[][] states = { new[] { AResource.Attribute.StateEnabled }, new[] { -AResource.Attribute.StateEnabled } };
+			int[] colors = { Control.TextColors.GetColorForState(states[0], Element.TextColor.ToAndroid()), Element.DisabledTextColor.ToAndroid().ToArgb() };
+
+			var colorStateList = new ColorStateList(states, colors);
+
+			Control.SetTextColor(colorStateList);
```
Comment (paymicro) [inline Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs]: Looks like this method updates not only the `DisabledTextColor`, but also the `TextColor`.

Comment (emmanuelvanmeerbeeck3factr) [reply Xamarin.Forms.Platform.Android/AppCompat/ButtonRenderer.cs]: I'm retrieving the current `TextColor`, and I set it as the `Enabled` text color. So in fact it updates it, but it doesn't set a new value, it just updates it to the same value. I don't know if I can do it another way. :-)

File: Xamarin.Forms.Core/Button.cs (hunk 1 above)
Comment (emmanuelvanmeerbeeck3factr) [reply Xamarin.Forms.Core/Button.cs]: This is indeed a more type-safe way of typing this, but I'm just following how it is already done for all the other properties, to stay consistent. :-)

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Button.cs]: it's ok to use `nameof()` even if the other ones do not. We do not fix code that works, as it defeat history, so new and old code can differ...

**Comment by samhouts** [comment]
@emmanuelvanmeerbeeck3factr Are you able to address the feedback? Thanks!

**Comment by samhouts** [comment]
@emmanuelvanmeerbeeck3factr Please reopen if you're able to work on this again. Thanks!

**Comment by emmanuelvanmeerbeeck3factr** [comment]
@samhouts  I finally could address the change requests. But I'm not able to reopen this pull request. Do I need to create a new one or am I not looking in the right place? I nowhere see the option to reopen it. 
@StephaneDelcroix  I have no experience with other platforms than Android/iOS, so I'm not able to implement this property for the other platforms. But when I look for UWP, I can't find a possibility to implement this functionality.

**Comment by StephaneDelcroix** [comment]
reopened

**Comment by jsuarezruiz** [comment]
@emmanuelvanmeerbeeck3factr The PR has been reopened. Are you able to address the feedback (and conflicts)?. Let us know if you need anything. Thanks!

**Comment by emmanuelvanmeerbeeck3factr** [comment]
> @emmanuelvanmeerbeeck3factr The PR has been reopened. Are you able to address the feedback (and conflicts)?. Let us know if you need anything. Thanks!

@jsuarezruiz I solved the conflicts and I addressed the change requests! Thanks.

File: Xamarin.Forms.Material.Android/MaterialButtonRenderer.cs [hunk 3]
```diff
@@ -284,8 +284,8 @@ void UpdatePrimaryColors()
 			else
 				text = textColor.ToAndroid();
 
-			// apply
-			SetTextColor(MaterialColors.CreateButtonTextColors(background, text));
+			// apply			
+			SetTextColor(MaterialColors.CreateButtonTextColors(background, text, Element.DisabledTextColor.ToAndroid()));
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Material.Android/MaterialButtonRenderer.cs]: Should checked if DisabledTextColor is the default color, and in that case use a disable text color from MaterialColors.

File: Xamarin.Forms.Platform.iOS/Renderers/ButtonElementManager.cs [hunk 4]
```diff
@@ -85,5 +85,11 @@ internal static void OnButtonTouchUpInside(IButtonController element)
 			element?.SendReleased();
 			element?.SendClicked();
 		}
+
+		public static void UpdateDisabledTextColor(Button button, UIButton nativeButton)
+		{
+			var title = new NSAttributedString(nativeButton.CurrentTitle, new UIStringAttributes { ForegroundColor = button.DisabledTextColor.ToUIColor() });
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Platform.iOS/Renderers/ButtonElementManager.cs]: `var title = new NSAttributedString(nativeButton.CurrentTitle ?? string.Empty, new UIStringAttributes { ForegroundColor = button.DisabledTextColor.ToUIColor() });`

This change fix the iOS issue:
Captura de pantalla 2019-11-06 a las 9 26 25

**Comment by samhouts** [comment]
@emmanuelvanmeerbeeck3factr Did you see @jsuarezruiz's latest comment? Thanks!

**Comment by emmanuelvanmeerbeeck3factr** [comment]
> @emmanuelvanmeerbeeck3factr Did you see @jsuarezruiz's latest comment? Thanks!

Yes I saw it now! Will take this up as soon as possible!

**Comment by PureWeen** [comment]
@jsuarezruiz Should we add a new property?

State based colors should always just use the VSM shouldn't they?

**Comment by dylanberry** [comment]
> @jsuarezruiz Should we add a new property?
> 
> State based colors should always just use the VSM shouldn't they?

VSM is a lot of markup compared to a property. Just my $0.02 :)

**Comment by ChaseFlorell** [comment]
I used vsm for button state a couple months ago. I could NOT get it looking the way I wanted, full stop. Also, the VSM markup that got me "close" was ~30 lines long.

**Comment by ChaseFlorell** [comment]
I'm cool with opinionated defaults, but please give me control over every pixel.

**Comment by Tommigun1980** [comment]
Hi. I need to change the text color on a disabled button, as I have a gray overlay on top of it the disabled button's text is indistinguishable. Will this pull request make it in soon?

Thanks.

**Comment by jfversluis** [comment]
Closing this as we won't be adding anything new to Forms

**Pull request closed without merging**